File: TaskManager/src/main/java/org.example/CommandManager.java
package org.example;

    import java.util.Stack;

    public class CommandManager {
        private Stack<Command> executedCommands = new Stack<>();
        private Stack<Command> undoneCommands = new Stack<>();

        public void executeCommand(Command command) {
            command.execute();
            executedCommands.push(command);
            undoneCommands.clear();
        }

        public void undoCommand() {
            if (!executedCommands.isEmpty()) {
                Command command = executedCommands.pop();
                command.undo();
                undoneCommands.push(command);
            }
        }

        public void redoCommand() {
            if (!undoneCommands.isEmpty()) {
                Command command = undoneCommands.pop();
                command.execute();
                executedCommands.push(command);
            }
        }
    }

File: TaskManager/LabFinal/LabFinal/CommandManager.cs
namespace LabFinal;

    public class CommandManager
    {
        private readonly Stack<ICommand> _executedCommands = new Stack<ICommand>();
        private readonly Stack<ICommand> _undoneCommands = new Stack<ICommand>();

        public void ExecuteCommand(ICommand command)
        {
            command.Execute();
            _executedCommands.Push(command);
            _undoneCommands.Clear();
        }

        public void Undo()
        {
            if (_executedCommands.Count > 0)
            {
                var command = _executedCommands.Pop();
                command.Undo();
                _undoneCommands.Push(command);
            }
        }

        public void Redo()
        {
            if (_undoneCommands.Count > 0)
            {
                var command = _undoneCommands.Pop();
                command.Execute();
                _executedCommands.Push(command);
            }
        }

        public IEnumerable<CommandLog> GetCommandHistory()
        {
            return _executedCommands.Select(c => c.GetLog());
        }
    }

